### PR TITLE
[MXNET-810] Add support for more req patterns for bilinear sampler backward

### LINKED ISSUE #8866: src/operator/./bilinear_sampler-inl.h:105: Have not implemented the data req combinations! gdata_req=0 ggrid_req=1
iam in the path of semantic segmentation. I want to warp encoder layer results of previous frame to current frame using optical flow based CNN. Basically i had key_img, cur_img and feat_key to warp the optical flow results and pass to the decoder.

Problem with mx.sym.BilinearSampler symbol code. KIndly provide the inputs.

Below is my symbol.

def get_cur_train_symbol( num_classes):
data_cur = mx.sym.Variable(name="data_cur")
data_key = mx.sym.Variable(name="data_key")
conv_feat = mx.sym.Variable(name="feat_key")
# shared convolutional layers
flow, scale_map = get_flownet(data_cur, data_key)
flow_grid = mx.sym.GridGenerator(data=flow, transform_type='warp', name='flow_grid')
conv_feat = mx.sym.BilinearSampler(data=conv_feat, grid=flow_grid, name='warping_feat')
conv_feat = conv_feat * scale_map
#conv_feats = mx.sym.SliceChannel(conv_feat, axis=1, num_outputs=2)

    fc6_bias = mx.symbol.Variable('fc6_bias', lr_mult=2.0)
    fc6_weight = mx.symbol.Variable('fc6_weight', lr_mult=1.0)
    fc6 = mx.symbol.Convolution(data=conv_feat, kernel=(1, 1), pad=(0, 0), num_filter=1024, name="fc6", bias=fc6_bias, weight=fc6_weight,workspace=workspace)
    relu_fc6 = mx.sym.Activation(data=fc6, act_type='relu', name='relu_fc6')
    score_bias = mx.symbol.Variable('score_bias', lr_mult=2.0)
    score_weight = mx.symbol.Variable('score_weight', lr_mult=1.0)
    score = mx.symbol.Convolution(data=relu_fc6, kernel=(1, 1), pad=(0, 0), num_filter=num_classes, name="score", bias=score_bias,weight=score_weight, workspace=workspace)

    upsampling = mx.symbol.Deconvolution(
        data=score, num_filter=num_classes, kernel=(32, 32), stride=(16, 16), num_group=num_classes, no_bias=True,
        name='upsampling', attr={'lr_mult': '0.0'}, workspace=workspace)

    croped_score = mx.symbol.Crop(*[upsampling, data_cur], offset=(8, 8), name='croped_score')
    softmax = mx.symbol.SoftmaxOutput(data=croped_score, normalization='valid', multi_output=True, use_ignore=True,ignore_label=255, name="softmaxoutput")
    return softmax

### PR DESCRIPTION
**Pull request by haojin2**
## Description ##
Fix #8866.

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Support more req type patterns for backward of bilinear sampler
- [x] Corresponding unit tests

## Comments ##
Added test case also passed flakiness check
```
MXNET_TEST_COUNT=10000 nosetests -s --verbose tests/python/gpu/test_operator_gpu.py:test_bilinear_sampler_versions

[INFO] Setting module np/mx/python random seeds, use MXNET_MODULE_SEED=583854465 to reproduce.
test_operator_gpu.test_bilinear_sampler_versions ... ok

----------------------------------------------------------------------
Ran 1 test in 539.629s

OK
```

### CODE REVIEW

**Comment by haojin2** [comment]
@eric-haibin-lin @reminisce @piiswrong @anirudh2290

File: tests/python/gpu/test_operator_gpu.py [hunk 1]
```diff
@@ -1975,6 +1975,19 @@ def test_bilinear_sampler_versions():
         assert_almost_equal(exe_list[ref_idx].grad_dict['data'].asnumpy(), data_grad + data_initial_grad, rtol=1e-3, atol=1e-5)
         assert_almost_equal(exe_list[ref_idx].grad_dict['grid'].asnumpy(), grid_grad + grid_initial_grad, rtol=1e-3, atol=1e-5)
 
+        req_dict = {'data' : 'null', 'grid' : 'write'}
```
Comment (anirudh2290) [inline tests/python/gpu/test_operator_gpu.py]: not testing req1 != Null ?

File: src/operator/mxnet_op.h [hunk 2]
```diff
@@ -111,6 +111,33 @@ inline int get_num_threads<cpu>(const int N) {
   }
 
 
+/*! \brief operator request type switch */
+#define MXNET_REQ_TYPE_SWITCH(req, ReqType, ...)  \
```
Comment (eric-haibin-lin) [inline src/operator/mxnet_op.h]: Why is this added?

File: tests/python/gpu/test_operator_gpu.py (hunk 1 above)
Comment (haojin2) [reply tests/python/gpu/test_operator_gpu.py]: Well the actual use case reported on the issues is this one, but maybe I should add the other one even if it's not a reported use case at all.

Comment (haojin2) [reply tests/python/gpu/test_operator_gpu.py]: Added extra test case as per your review, should be good for merge.

**Comment by haojin2** [comment]
@anirudh2290 @eric-haibin-lin Should be good for merge?
